# Exercise 1.2, Exploring a Pandas Dataframe
## Teresa Ferrill, September 14, 2025

Download the Video Game Sales with Ratings dataset from this link: Video Game Sales with Ratings, https://www.kaggle.com/rush4ratio/video-game-sales-with-ratings. 

-----------------------------------------------------------------------------------------------------------------
*1.   Load the dataset as a Pandas data frame.*

In [98]:
# Load pandas Library
import pandas as pd

# Load Data
# CSV file located at root folder for this notebook
dataframe = pd.read_csv('Video_Games_Sales_as_at_22_Dec_2016.csv') 

-----------------------------------------------------------------------------------------------------------------
*2. Display the first ten rows of data.*

In [99]:
# View first 10 rows of the dataframe
dataframe.head(10)

,Name,Platform,Year_of_Release,Genre,Publisher,NA_Sales,EU_Sales,JP_Sales,Other_Sales,Global_Sales,Critic_Score,Critic_Count,User_Score,User_Count,Developer,Rating
0,Wii Sports,Wii,2006.0,Sports,Nintendo,41.36,28.96,3.77,8.45,82.53,76.0,51.0,8,322.0,Nintendo,E
1,Super Mario Bros.,NES,1985.0,Platform,Nintendo,29.08,3.58,6.81,0.77,40.24,NaN,NaN,NaN,NaN,NaN,NaN
2,Mario Kart Wii,Wii,2008.0,Racing,Nintendo,15.68,12.76,3.79,3.29,35.52,82.0,73.0,8.3,709.0,Nintendo,E
3,Wii Sports Resort,Wii,2009.0,Sports,Nintendo,15.61,10.93,3.28,2.95,32.77,80.0,73.0,8,192.0,Nintendo,E
4,Pokemon Red/Pokemon Blue,GB,1996.0,Role-Playing,Nintendo,11.27,8.89,10.22,1.00,31.37,NaN,NaN,NaN,NaN,NaN,NaN
5,Tetris,GB,1989.0,Puzzle,Nintendo,23.20,2.26,4.22,0.58,30.26,NaN,NaN,NaN,NaN,NaN,NaN
6,New Super Mario Bros.,DS,2006.0,Platform,Nintendo,11.28,9.14,6.50,2.88,29.80,89.0,65.0,8.5,431.0,Nintendo,E
7,Wii Play,Wii,2006.0,Misc,Nintendo,13.96,9.18,2.93,2.84,28.92,58.0,41.0,6.6,129.0,Nintendo,E
8,New Super Mario Bros. Wii,Wii,2009.0,Platform,Nintendo,14.44,6.94,4.70,2.24,28.32,87.0,80.0,8.4,594.0,Nintendo,E
9,Duck Hunt,NES,1984.0,Shooter,Nintendo,26.93,0.63,0.28,0.47,28.31,NaN,NaN,NaN,NaN,NaN,NaN


-----------------------------------------------------------------------------------------------------------------
*3. Find the dimensions (number of rows and columns) in the data frame. What do these two numbers represent in the context of the data?*

In [102]:
# Load pandas Library
import pandas as pd

# Load Data
# CSV file located at root folder for this notebook
dataframe = pd.read_csv('Video_Games_Sales_as_at_22_Dec_2016.csv') 

# Show dimensions
# 'shape' returns the number of rows and the number of columns
print("Dataframe Dimensions - (number-of-rows, number-of-columns)")
dataframe.shape

Dataframe Dimensions - (number-of-rows, number-of-columns)


(16719, 16)

The first number (16719) is the number of rows and the second number (16) is the number of columns.  
Each row represents a single observation/record/data point within the dataset.  
Each column represents a specific attribute/feature/variable for each observation.

-----------------------------------------------------------------------------------------------------------------
*4. Find the top five games by critic score.*

In [103]:
# Load pandas Library
import pandas as pd

# Load Data
# CSV file located at root folder for this notebook
dataframe = pd.read_csv('Video_Games_Sales_as_at_22_Dec_2016.csv')

# Sort dataframe by critic score (descending to list highest numbers first), show 5 rows
# Added 'ascending=False' for readability - to list the highest Critic_Score first, followed by the next highest value
dataframe.sort_values(by='Critic_Score', ascending=False).head(5)

,Name,Platform,Year_of_Release,Genre,Publisher,NA_Sales,EU_Sales,JP_Sales,Other_Sales,Global_Sales,Critic_Score,Critic_Count,User_Score,User_Count,Developer,Rating
57,Grand Theft Auto IV,PS3,2008.0,Action,Take-Two Interactive,4.76,3.69,0.44,1.61,10.50,98.0,64.0,7.5,2833.0,Rockstar North,M
5350,SoulCalibur,DC,1999.0,Fighting,Namco Bandai Games,0.00,0.00,0.34,0.00,0.34,98.0,24.0,8.8,200.0,Namco,T
227,Tony Hawk's Pro Skater 2,PS,2000.0,Sports,Activision,3.05,1.41,0.02,0.20,4.68,98.0,19.0,7.7,299.0,Neversoft Entertainment,T
51,Grand Theft Auto IV,X360,2008.0,Action,Take-Two Interactive,6.76,3.07,0.14,1.03,11.01,98.0,86.0,7.9,2951.0,Rockstar North,M
165,Grand Theft Auto V,XOne,2014.0,Action,Take-Two Interactive,2.81,2.19,0.00,0.47,5.48,97.0,14.0,7.9,764.0,Rockstar North,M


-----------------------------------------------------------------------------------------------------------------
*5. Find the number of video games in the data frame in each genre.*

In [110]:
# Load pandas Library
import pandas as pd

# Load Data
# CSV file located at root folder for this notebook
dataframe = pd.read_csv('Video_Games_Sales_as_at_22_Dec_2016.csv') 

# Get the count of games in each genre
genre_counts = dataframe['Genre'].value_counts()

# Group by 'Genre' and count the number of games in each genre
genre_counts = dataframe.groupby('Genre').size()

# Sort the results by genre (alphabetically by default) for readability
sorted_genre_counts = genre_counts.sort_index()

print("Number of video games in each genre, sorted by genre:")
print(sorted_genre_counts)

Number of video games in each genre, sorted by genre:
Genre
Action          3370
Adventure       1303
Fighting         849
Misc            1750
Platform         888
Puzzle           580
Racing          1249
Role-Playing    1500
Shooter         1323
Simulation       874
Sports          2348
Strategy         683
dtype: int64


-----------------------------------------------------------------------------------------------------------------
*6. Find the first five games in the data frame on the SNES platform.*

In [111]:
# Load pandas Library
import pandas as pd

# Load Data
# CSV file located at root folder for this notebook
dataframe = pd.read_csv('Video_Games_Sales_as_at_22_Dec_2016.csv') 

# Filter for SNES platform games
snes_games = dataframe[dataframe['Platform'] == 'SNES']

# Print first five games in the SNES dataframe
snes_games.head(5)

,Name,Platform,Year_of_Release,Genre,Publisher,NA_Sales,EU_Sales,JP_Sales,Other_Sales,Global_Sales,Critic_Score,Critic_Count,User_Score,User_Count,Developer,Rating
18,Super Mario World,SNES,1990.0,Platform,Nintendo,12.78,3.75,3.54,0.55,20.61,NaN,NaN,NaN,NaN,NaN,NaN
56,Super Mario All-Stars,SNES,1993.0,Platform,Nintendo,5.99,2.15,2.12,0.29,10.55,NaN,NaN,NaN,NaN,NaN,NaN
71,Donkey Kong Country,SNES,1994.0,Platform,Nintendo,4.36,1.71,3.00,0.23,9.30,NaN,NaN,NaN,NaN,NaN,NaN
76,Super Mario Kart,SNES,1992.0,Racing,Nintendo,3.54,1.24,3.81,0.18,8.76,NaN,NaN,NaN,NaN,NaN,NaN
137,Street Fighter II: The World Warrior,SNES,1992.0,Fighting,Capcom,2.47,0.83,2.87,0.12,6.30,NaN,NaN,NaN,NaN,NaN,NaN


-----------------------------------------------------------------------------------------------------------------
*7. Find the five publishers with the highest total global sales. Note: You will need to calculate the total global sales for each publisher to do this.*

In [112]:
# Load pandas Library
import pandas as pd

# Load Data
# CSV file located at root folder for this notebook
dataframe = pd.read_csv('Video_Games_Sales_as_at_22_Dec_2016.csv') 

# Calculate total global sales for each publisher
publisher_sales = dataframe.groupby('Publisher')['Global_Sales'].sum()

# Sort in descending order and get the top 5
top_5_publishers = publisher_sales.sort_values(ascending=False).head(5)

print("Top 5 publishers with the highest total global sales:")
print(top_5_publishers)

Top 5 publishers with the highest total global sales:
Publisher
Nintendo                       1788.81
Electronic Arts                1116.96
Activision                      731.16
Sony Computer Entertainment     606.48
Ubisoft                         471.61
Name: Global_Sales, dtype: float64


-----------------------------------------------------------------------------------------------------------------
*8.Create a new column in the data frame that calculates the percentage of global sales from North America. Display the first five rows of the new data frame.*

In [113]:
# Load pandas Library
import pandas as pd

# Load Data
# CSV file located at root folder for this notebook
dataframe = pd.read_csv('Video_Games_Sales_as_at_22_Dec_2016.csv') 

# Create a new column that includes 
dataframe['NA_Percentage'] = ((dataframe['NA_Sales'] / dataframe['Global_Sales']) * 100)
dataframe.head(5)

,Name,Platform,Year_of_Release,Genre,Publisher,NA_Sales,EU_Sales,JP_Sales,Other_Sales,Global_Sales,Critic_Score,Critic_Count,User_Score,User_Count,Developer,Rating,NA_Percentage
0,Wii Sports,Wii,2006.0,Sports,Nintendo,41.36,28.96,3.77,8.45,82.53,76.0,51.0,8,322.0,Nintendo,E,50.115110
1,Super Mario Bros.,NES,1985.0,Platform,Nintendo,29.08,3.58,6.81,0.77,40.24,NaN,NaN,NaN,NaN,NaN,NaN,72.266402
2,Mario Kart Wii,Wii,2008.0,Racing,Nintendo,15.68,12.76,3.79,3.29,35.52,82.0,73.0,8.3,709.0,Nintendo,E,44.144144
3,Wii Sports Resort,Wii,2009.0,Sports,Nintendo,15.61,10.93,3.28,2.95,32.77,80.0,73.0,8,192.0,Nintendo,E,47.635032
4,Pokemon Red/Pokemon Blue,GB,1996.0,Role-Playing,Nintendo,11.27,8.89,10.22,1.00,31.37,NaN,NaN,NaN,NaN,NaN,NaN,35.926044


-----------------------------------------------------------------------------------------------------------------
*9. Find the number NaN entries (missing data values) in each column.*

In [115]:
# Load pandas Library
import pandas as pd

# Load Data
# CSV file located at root folder for this notebook
dataframe = pd.read_csv('Video_Games_Sales_as_at_22_Dec_2016.csv') 

# Count the number of missing (null or NaN) values in each column of the DataFrame
# Print the results
print(dataframe.isnull().sum())

Name                  2
Platform              0
Year_of_Release     269
Genre                 2
Publisher            54
NA_Sales              0
EU_Sales              0
JP_Sales              0
Other_Sales           0
Global_Sales          0
Critic_Score       8582
Critic_Count       8582
User_Score         6704
User_Count         9129
Developer          6623
Rating             6769
dtype: int64


-----------------------------------------------------------------------------------------------------------------
*10. Try to calculate the median user score of all the video games. You will likely run into an error because some of the user score entries are a non-numerical string that cannot be converted to a float. Find and replace this string with NaN and then calculate the median. Then, replace all NaN entries in the user score column with the median value.*

In [116]:
# Load pandas Library
import pandas as pd

# Load Data
# CSV file located at root folder for this notebook
dataframe = pd.read_csv('Video_Games_Sales_as_at_22_Dec_2016.csv') 

# Print selected rows from the dataset to include rows with numbers, letters, and null values
# Reivew of the CSV file shows the #119 (Zumba Fitness) row has 'tbd' for the User_Score 
#   while #1 and #201 (Super Mario Bros.) rows have null values
dataframe[dataframe['Name'].isin(['Zumba Fitness','Super Mario Bros.'])] 

,Name,Platform,Year_of_Release,Genre,Publisher,NA_Sales,EU_Sales,JP_Sales,Other_Sales,Global_Sales,Critic_Score,Critic_Count,User_Score,User_Count,Developer,Rating
1,Super Mario Bros.,NES,1985.0,Platform,Nintendo,29.08,3.58,6.81,0.77,40.24,NaN,NaN,NaN,NaN,NaN,NaN
119,Zumba Fitness,Wii,2010.0,Sports,505 Games,3.45,2.59,0.00,0.66,6.71,NaN,NaN,tbd,NaN,"Pipeworks Software, Inc.",E
201,Super Mario Bros.,GB,1999.0,Platform,Nintendo,3.40,1.30,0.15,0.22,5.07,NaN,NaN,NaN,NaN,NaN,NaN
669,Zumba Fitness,X360,2010.0,Sports,505 Games,1.74,0.45,0.00,0.18,2.37,42.0,10.0,5.5,16.0,"Pipeworks Software, Inc.",E
3516,Zumba Fitness,PS3,2010.0,Sports,505 Games,0.28,0.21,0.00,0.08,0.57,NaN,NaN,3.8,4.0,"Pipeworks Software, Inc.",E


In [117]:
# Convert the User_Score column type to numeric
# Handle any conversion errors by coercion (errors='coerce'), which replaces the non-numeric value with NaN
dataframe['User_Score'] = pd.to_numeric(dataframe['User_Score'], errors='coerce')

# Calculate the median of the User_Score column
median_user_score = dataframe['User_Score'].median()

# Print out the median value
print("Median_User_Score value")
print(median_user_score)

Median_User_Score value
7.5


In [118]:
# Print selected dataframe rows - the 'tbd' value has been replaced with a NaN value
print("Dataframe with NaN for non-numeric User_Score values")
dataframe[dataframe['Name'].isin(['Zumba Fitness','Super Mario Bros.'])]

Dataframe with NaN for non-numeric User_Score values


,Name,Platform,Year_of_Release,Genre,Publisher,NA_Sales,EU_Sales,JP_Sales,Other_Sales,Global_Sales,Critic_Score,Critic_Count,User_Score,User_Count,Developer,Rating
1,Super Mario Bros.,NES,1985.0,Platform,Nintendo,29.08,3.58,6.81,0.77,40.24,NaN,NaN,NaN,NaN,NaN,NaN
119,Zumba Fitness,Wii,2010.0,Sports,505 Games,3.45,2.59,0.00,0.66,6.71,NaN,NaN,NaN,NaN,"Pipeworks Software, Inc.",E
201,Super Mario Bros.,GB,1999.0,Platform,Nintendo,3.40,1.30,0.15,0.22,5.07,NaN,NaN,NaN,NaN,NaN,NaN
669,Zumba Fitness,X360,2010.0,Sports,505 Games,1.74,0.45,0.00,0.18,2.37,42.0,10.0,5.5,16.0,"Pipeworks Software, Inc.",E
3516,Zumba Fitness,PS3,2010.0,Sports,505 Games,0.28,0.21,0.00,0.08,0.57,NaN,NaN,3.8,4.0,"Pipeworks Software, Inc.",E


In [119]:
# Replace NaN entries in User_Score with the calculated Median_User_Score value
dataframe['User_Score'] = dataframe['User_Score'].fillna(median_user_score)

# Print selected rows - the NaN values are replaced with the median_user_score value
print("Dataframe with NaN values replaced with median_user_score")
dataframe[dataframe['Name'].isin(['Zumba Fitness','Super Mario Bros.'])]

Dataframe with NaN values replaced with median_user_score


,Name,Platform,Year_of_Release,Genre,Publisher,NA_Sales,EU_Sales,JP_Sales,Other_Sales,Global_Sales,Critic_Score,Critic_Count,User_Score,User_Count,Developer,Rating
1,Super Mario Bros.,NES,1985.0,Platform,Nintendo,29.08,3.58,6.81,0.77,40.24,NaN,NaN,7.5,NaN,NaN,NaN
119,Zumba Fitness,Wii,2010.0,Sports,505 Games,3.45,2.59,0.00,0.66,6.71,NaN,NaN,7.5,NaN,"Pipeworks Software, Inc.",E
201,Super Mario Bros.,GB,1999.0,Platform,Nintendo,3.40,1.30,0.15,0.22,5.07,NaN,NaN,7.5,NaN,NaN,NaN
669,Zumba Fitness,X360,2010.0,Sports,505 Games,1.74,0.45,0.00,0.18,2.37,42.0,10.0,5.5,16.0,"Pipeworks Software, Inc.",E
3516,Zumba Fitness,PS3,2010.0,Sports,505 Games,0.28,0.21,0.00,0.08,0.57,NaN,NaN,3.8,4.0,"Pipeworks Software, Inc.",E
